lib/src/front_end/piece_factory.dart

```diff
@@ -427,6 +431,34 @@ mixin PieceFactory implements CommentWriter {
         isValueDelimited: rightHandSide.isDelimited));
   }
 
+  /// Writes the condition and updaters part of for a [ForParts] after the
+  /// subclass's initializer clause has been written.
+  void finishForParts(ForParts forLoopParts, DelimitedListBuilder partsList) {
+    token(forLoopParts.leftSeparator);
+    writer.split();
+    partsList.add(writer.pop());
+
+    // The condition clause.
+    if (forLoopParts.condition case var conditionExpression?) {
+      partsList.addCommentsBefore(conditionExpression.beginToken);
+      visit(conditionExpression);
+    } else {
+      partsList.addCommentsBefore(forLoopParts.rightSeparator);
+    }
+
+    token(forLoopParts.rightSeparator);
+    writer.split();
+    partsList.add(writer.pop());
```

Comment:
Just a thought for later: The pattern of `pop` and `add` seems to happen a lot. It's also a little confusing to read. Maybe change the name or group the API at some point in the future?

Reply:
Agreed totally. I'm working on a refresh of the API to make this better.